src/kbmod/search/raw_image_eigen.cpp

```diff
@@ -0,0 +1,767 @@
+            num_unmasked += 1;
+          }
+        }
+
+        if (num_unmasked > 0) {
+          // how is it possible that this optimization ever pays off?
+          std::sort(pix_array.begin(), pix_array.begin() + num_unmasked);
+
+          // If we have an even number of elements, take the mean of the two
+          // middle ones.
+          int median_ind = num_unmasked / 2;
+          if (num_unmasked % 2 == 0) {
+            float ave_middle = (pix_array[median_ind] + pix_array[median_ind - 1]) / 2.0;
+            result.set_pixel(x, y, ave_middle);
+          } else {
+            result.set_pixel(x, y, pix_array[median_ind]);
+          }
+        } else {
+          // We use a 0.0 value if there is no data to allow for visualization
+          // and value based filtering.
+          result.set_pixel(x, y, 0.0);
+        }
+      }
+    }
+
+    return result;
+  }
+
+
+  RawImageEigen create_summed_image_eigen(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+    int width = images[0].get_width();
+    int height = images[0].get_height();
+    RawImageEigen result(width, height);
+    for (auto& img : images){
+      result.image += img.image;
+    }
+    return result;
+  }
+
+
+  RawImageEigen create_mean_image_eigen(const std::vector<RawImageEigen>& images) {
+    int num_images = images.size();
+    assert(num_images > 0);
+
+    auto result = create_summed_image_eigen(images);
+    result.image.array() /= num_images;
+    result.image = (result.image.array() == NO_DATA).select(0, result.image);
+    return result;
+  }
+
+
+#ifdef Py_PYTHON_H
+  static void raw_image_eigen_bindings(py::module &m) {
+    using rie = search::RawImageEigen;
+
+    py::class_<rie>(m, "RawImageEigen")
+      .def(py::init<>())
+      .def(py::init<search::Image&, double>(),
+           py::arg("img").noconvert(true), py::arg("obstime"))
+      .def(py::init<unsigned, unsigned, float, double>(),
+           py::arg("w"), py::arg("h"), py::arg("value")=0.0f, py::arg("obs_time")=-1.0d)
+      .def("height", &rie::get_height)
+      .def("width", &rie::get_width)
+      //.def_property("data",
+      //              /*getter=*/ [](rie& cls) { return ImageRef(cls.data); },
+      //              /*setter=*/ [](rie& cls, ImageRef value) { cls.array = value; })
+      .def_property("obstime", &rie::get_obstime, &rie::set_obstime)
+      .def("isclose", &rie::isclose)
+      .def("compute_bounds", &rie::compute_bounds)
+      .def("find_peak", &rie::find_peak)
+      .def("find_central_moments", &rie::find_central_moments)
+      .def("create_stamp", &rie::create_stamp)
+      .def("interpolate", &rie::interpolate)
+      .def("interpolate2", &rie::interpolate2)
+      .def("interpolated_add", &rie::interpolated_add)
+      // I really need to confirm whether or not references make a copy
+      // or behave like Eigen::Ref counterparts - we might want to use more of
```

Comment:
Returning references in C++ will not make a copy. Returning references in pybind11 shouldn't if you set `return_value_policy::reference_internal`

Reply:
As pointed out in the shared notes, it's not always obvious what type expressions evaluate to. Care must be taken to make sure `pybind11` doesn't cast an expression to a `numpy.array` by constructing `Image` instead of an `ImageRef` or `Image&`.